ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,914 @@
+        generate_for.push(cand.clone());
+        assert!(
+            generate_commitments_for(
+                SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDlog(pk1.clone())),
+                generate_for.as_slice(),
+            )
+            .hints
+            .is_empty(),
+            "{}",
+            "{}"
+        );
+        assert!(
+            generate_commitments_for(cand.clone(), generate_for.as_slice())
+                .hints
+                .is_empty(),
+            "{}",
+            "{}"
+        );
+        generate_for.clear();
+        generate_for.push(SigmaBoolean::ProofOfKnowledge(
+            SigmaProofOfKnowledgeTree::ProveDlog(pk1.clone()),
+        ));
+
+        assert!(
+            !generate_commitments_for(
+                SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDlog(pk1.clone())),
+                generate_for.as_slice(),
+            )
+            .hints
+            .is_empty(),
+            "{}",
+            "{}"
+        );
+        generate_for.clear();
+        generate_for.push(SigmaBoolean::ProofOfKnowledge(
+            SigmaProofOfKnowledgeTree::ProveDlog(pk1.clone()),
+        ));
+        let mut bag = generate_commitments_for(
+            SigmaBoolean::ProofOfKnowledge(SigmaProofOfKnowledgeTree::ProveDlog(pk1)),
+            generate_for.as_slice(),
+        );
+        assert!(!bag.hints.is_empty(), "{}", "{}");
+        let mut hint = bag.hints[0].clone();
+        let mut a: Option<FirstProverMessage> = None;
+        let mut r: Option<Scalar> = None;
+        if let Hint::CommitmentHint(CommitmentHint::RealCommitment(comm)) = hint {
+            assert_eq!(comm.position, NodePosition::crypto_tree_prefix());
+            a = Some(comm.commitment);
+        }
+        hint = bag.hints[1].clone();
+        if let Hint::CommitmentHint(CommitmentHint::OwnCommitment(comm)) = hint {
+            assert_eq!(comm.position, NodePosition::crypto_tree_prefix());
+            r = Some(comm.secret_randomness);
+        }
+        let g_to_r = dlog_group::exponentiate(&dlog_group::generator(), &r.unwrap());
+        assert_eq!(
+            FirstProverMessage::FirstDlogProverMessage(g_to_r.into()),
+            a.unwrap()
+        );
+
+        bag = generate_commitments_for(cand, generate_for.as_slice());
+        assert_eq!(bag.hints.len(), 2);
+        hint = bag.hints[0].clone();
+        if let Hint::CommitmentHint(CommitmentHint::RealCommitment(comm)) = hint {
+            assert_eq!(
+                comm.position,
+                NodePosition {
+                    positions: vec![0, 0]
+                }
+            );
+        }
+        hint = bag.hints[1].clone();
+        if let Hint::CommitmentHint(CommitmentHint::OwnCommitment(comm)) = hint {
+            assert_eq!(
+                comm.position,
+                NodePosition {
+                    positions: vec![0, 0]
+                }
+            );
+        }
```

Comment:
DH case is missing according to https://github.com/ScorexFoundation/sigmastate-interpreter/blob/795c176debfe105646a7deef3f01bd537b79c81d/sigmastate/src/test/scala/sigmastate/utxo/ProverSpecification.scala#L49-L57

Reply:
My bad i changed them in new commit.